### ISSUE TITLE
photon-ml really can support large-scale ID class features? 

### ISSUE THREAD

**Issue by clcron** [linkedin/photon-ml#197]
I found some comments at class com.linkedin.photon.ml.util.DefaultIndexMap

 * Use the default system HashMap to construct an index map, highly inefficient in terms of memory usage; but easier to
  * handle. Recommended for small feature space cases (<= 200k).

200K can be considered a large scale ?

**Comment by clcron** [comment]
when my feature space cases increase to 20000k ~ photo-ml will OOM ~ 
I think photo-ml need support "Model Slice Training"~so that photon-ml  can claim support large-scale

**Comment by XianXing** [comment]
Thanks for submitting the issue.

As you pointed out and documented in the code, ```com.linkedin.photon.ml.util.DefaultIndexMap``` is designed to handle small feature space cases where number of features <= 200k. In case you are interested, such limitation is due to the fact that the feature map (a map from feature raw string to feature index) is broadcasted from master to worker nodes and cached in memory in ```DefaultIndexMap```.

If we have more than 200k features (e.g., 20,000k in your case), ```com.linkedin.photon.ml.util. PalDBIndexMap``` might be a better choice than ```DefaultIndexMap```, which leverages a (distributed) Key-Value store to handle the feature map. With ```PalDBIndexMap```, Photon-ML should be able to handle billions features with proper setting of executor memory (e.g., > 5GB per executor) and sufficient number of executors given the size of the training data. Please let us (@timyitong, @joshvfleming) know if you need any help in setting up your model training job with ```PalDBIndexMap```. 

And since you mentioned large-scale id class features in the title of this issue, I wish to point out that sometimes using id class as features may end up with tens even hundreds billions of features (consider user id or item id as features), which sometimes can be computationally challenging. And that's why we introduced the GAME module in Photon-ML, a framework that is designed specifically to handle such id-level features with very high dimensional feature space. Please let us know if you are interested to try it out.

**Comment by XianXing** [comment]
@clcron Besides, I am not very familiar with the term "Model Slice Training" you mentioned, is it something similar to what've been done in MXNet for neural network? Anyway, just want to let you know that Photon-ML always welcomes external contributions :)

**Comment by clcron** [comment]
@XianXing forgive my ignorance and poor english~
Thanks for your prompt. I will try "PalDBIndexMap" that you mentioned ~
actually I mean implements data and model parallelism like [Downpour SGD](http://static.googleusercontent.com/media/research.google.com/zh-CN//archive/large_deep_networks_nips2012.pdf)  ~ 
:) thanks for your Reply~

**Comment by clcron** [comment]
@XianXing 
hi~ I found the initial model code 
 /**
   * Run the algorithm with the configured parameters on an input RDD of LabeledPoint entries.
   *
   * @param input A RDD of input labeled data points in the original scale
   * @param normalizationContext The normalization context
   * @return The learned generalized linear models of each regularization weight and iteration.
   */
` 
  def run(input: RDD[LabeledPoint], normalizationContext: NormalizationContext): GLM = {
    val numFeatures = input.first().features.size
    val initialWeight = Vector.zeros\[Double\](numFeatures)
    val initialModel = createModel(initialWeight, None)

    run(input, initialModel, normalizationContext)
  }`

the initialWeight  is a DenseVector which is a Array Implement ~ if the vector size is big~ could the DenseVector be the bottleneck ?

**Comment by zhenchuan** [comment]
The question is how much of your feature number? 
If it's really a lot, it'll be  a bottleneck.

A double is 8 bytes, assuming you have the features 10^9, it takes up 8G of the memory.
So much of the feature num is always due to the combination of ids (eg:user id)

In front of @XianXing said that you can use their GAME model to deal with the combination between ID features

PalDBIndexMap was introduced to  optimize for large training samples

**Comment by timyitong** [comment]
@clcron To add to @zhenchuan 's point, [PalDB](https://github.com/linkedin/PalDB) is storing data in an off heap fashion. Only the hash keys of the keys are kept in memory and original keys / values are stored offheap using ser/der. The length of the hashes could variate from 2 bytes to a large amount of bytes, but roughly speaking in the current photon-ml's use case, as long as `[the total number of unique features] x roughly [4~8] bytes` does not exceed your executor memory, the index map should work, regardless of how long the feature name string could be since it'll be stored offheap regardless.

**Comment by clcron** [comment]
Why not user SparseVector implement ~ may be can support more features?

**Comment by timyitong** [comment]
@clcron each data point might have sparse feature vectors, but a model vector is collecting information from all data points; even you use sparse vector, non-zero parameter percentage will finally reach to almost 100%, making sparse vector even worse in terms of space usage than dense vector (also need to store vector indices).

So in the end, some other tricks would be needed to overcome this limit. That's one of the motivation of introducing GAME. You can also check out [parameter server](https://github.com/dmlc/ps-lite) which is basically also distributing the parameter storage/operation part.

**Comment by clcron** [comment]
@timyitong  I don't think so , if use L1-regularization ~ may be the model vector can not reach 100% ~

**Comment by fastier-li** [comment]
@clcron - I am not sure what feature you are talking about. Can you be more specific? With more details, I can hopefully guide you to the right setting for GAME. Thanks!

**Comment by fastier-li** [comment]
@cicron - if I don't hear anything back by this week's end, I'll consider that the issue can be closed for now. We can always open a new one later. Thanks!
